File: crypto/src/main/java/io/oversec/one/crypto/encoding/pad/XCoderAndPadderFactory.java
package io.oversec.one.crypto.encoding.pad;

import android.content.Context;
import io.oversec.one.common.CoreContract;
import io.oversec.one.crypto.Issues;
import io.oversec.one.crypto.encoding.XCoderAndPadder;
import io.oversec.one.crypto.encoding.XCoderFactory;

import java.util.ArrayList;
import java.util.List;

public class XCoderAndPadderFactory {


    private static XCoderAndPadderFactory INSTANCE;

    private final Context mCtx;
    private final CoreContract mCore;
    private ArrayList<XCoderAndPadder> mAll = new ArrayList<>();
    private ArrayList<XCoderAndPadder> mSym;
    private ArrayList<XCoderAndPadder> mGpg;
    private ArrayList<XCoderAndPadder> mSymExcludeInvisible;
    private ArrayList<XCoderAndPadder> mGpgExcludeInvisible;
    private XCoderAndPadder mManualZeroWidthXcoder;

    public static synchronized XCoderAndPadderFactory getInstance(Context ctx) {
        if (INSTANCE == null) {
            INSTANCE = new XCoderAndPadderFactory(ctx.getApplicationContext());
        }
        return INSTANCE;
    }

    private XCoderAndPadderFactory(Context ctx) {
        mCtx = ctx;
        mCore = CoreContract.getInstance();
        reload();


    }

    public void reload() {
        mAll.clear();
        mManualZeroWidthXcoder = addZeroWidthXcoder(new ManualPadder(mCtx));
        CoreContract contract = CoreContract.getInstance();

        List<PadderContent> allFromDb = contract.getAllPaddersSorted();
        for (PadderContent pc : allFromDb) {
            addZeroWidthXcoder(new GutenbergPadder(mCtx, pc.getName(), pc.getContent()));
        }


        mSym = new ArrayList<>(mAll);
        mGpg = new ArrayList<>(mAll);
        mSymExcludeInvisible = new ArrayList<>();
        mGpgExcludeInvisible = new ArrayList<>();

        XCoderAndPadder l = new XCoderAndPadder(XCoderFactory.getInstance(mCtx)._Base64XCoder, null);
        mSym.add(0, l);
        mSymExcludeInvisible.add(0, l);
        mAll.add(l);

        XCoderAndPadder l2 = new XCoderAndPadder(XCoderFactory.getInstance(mCtx)._AsciiArmouredGpgXCoder, null);
        mGpg.add(0, l2);
        mGpgExcludeInvisible.add(0, l2);
        mAll.add(l2);
    }

    private XCoderAndPadder addZeroWidthXcoder(AbstractPadder padder) {
        XCoderAndPadder res = new XCoderAndPadder(XCoderFactory.getInstance(mCtx)._ZeroWidthXCoder, padder);
        mAll.add(res);
        return res;
    }

    public ArrayList<XCoderAndPadder> getSym(String packagename) {
        ArrayList<XCoderAndPadder> res = new ArrayList<>(Issues.cantHandleInvisibleEncoding(packagename) ? mSymExcludeInvisible : mSym);
        if (mCore.isDbSpreadInvisibleEncoding(packagename)) {
            //spreaded doesn't work with manual padding
            res.remove(mManualZeroWidthXcoder);
        }

        return res;
    }


    public ArrayList<XCoderAndPadder> getGpg(String packagename) {
        ArrayList<XCoderAndPadder> res = new ArrayList<>(Issues.cantHandleInvisibleEncoding(packagename) ? mGpgExcludeInvisible:mGpg);
        if (mCore.isDbSpreadInvisibleEncoding(packagename)) {
            //spreaded doesn't work with manual padding
            res.remove(mManualZeroWidthXcoder);
        }

        return res;
    }

    public XCoderAndPadder get(String coderId, String padderId) {

        for (XCoderAndPadder x : mAll) {
            if (x.getXcoder().getId().equals(coderId)) {
                if (padderId == null || x.getPadder() == null) {
                    return x;
                } else {
                    if (x.getPadder().getId().equals(padderId)) {
                        return x;
                    }
                }
            }
        }

        return null;
    }
}

File: crypto/src/main/java/io/oversec/one/crypto/encoding/pad/XCoderAndPadderFactory.kt
package io.oversec.one.crypto.encoding.pad

import android.annotation.SuppressLint
import android.content.Context
import io.oversec.one.common.CoreContract
import io.oversec.one.crypto.Issues
import io.oversec.one.crypto.encoding.XCoderAndPadder
import io.oversec.one.crypto.encoding.XCoderFactory
import java.util.ArrayList

class XCoderAndPadderFactory private constructor(private val mCtx: Context) {
    private val mCore = CoreContract.instance
    private var mAll: MutableList<XCoderAndPadder> = ArrayList(0)
    private var mSym: MutableList<XCoderAndPadder> = ArrayList(0)
    private var mGpg: MutableList<XCoderAndPadder> = ArrayList(0)
    private var mSymExcludeInvisible: MutableList<XCoderAndPadder> = ArrayList(0)
    private var mGpgExcludeInvisible: MutableList<XCoderAndPadder> = ArrayList(0)
    private var mManualZeroWidthXcoder: XCoderAndPadder? = null

    fun reload() {
        mAll.clear()
        mManualZeroWidthXcoder = addZeroWidthXcoder(ManualPadder(mCtx))
        val contract = CoreContract.instance

        val allFromDb = contract.allPaddersSorted
        for (pc in allFromDb) {
            addZeroWidthXcoder(GutenbergPadder(mCtx, pc.name, pc.content))
        }

        mSym = ArrayList(mAll)
        mGpg = ArrayList(mAll)
        mSymExcludeInvisible = ArrayList()
        mGpgExcludeInvisible = ArrayList()

        val l = XCoderAndPadder(XCoderFactory.getInstance(mCtx).base64XCoder, null)
        mSym.add(0, l)
        mSymExcludeInvisible!!.add(0, l)
        mAll.add(l)

        val l2 = XCoderAndPadder(XCoderFactory.getInstance(mCtx).asciiArmouredGpgXCoder, null)
        mGpg!!.add(0, l2)
        mGpgExcludeInvisible!!.add(0, l2)
        mAll.add(l2)
    }

    private fun addZeroWidthXcoder(padder: AbstractPadder): XCoderAndPadder {
        val res = XCoderAndPadder(XCoderFactory.getInstance(mCtx).zeroWidthXCoder, padder)
        mAll.add(res)
        return res
    }

    fun getSym(packagename: String): ArrayList<XCoderAndPadder> {
        val res =
            ArrayList(if (Issues.cantHandleInvisibleEncoding(packagename)) mSymExcludeInvisible else mSym)
        if (mCore.isDbSpreadInvisibleEncoding(packagename)) {
            //spreaded doesn't work with manual padding
            res.remove(mManualZeroWidthXcoder)
        }

        return res
    }


    fun getGpg(packagename: String): ArrayList<XCoderAndPadder> {
        val res =
            ArrayList(if (Issues.cantHandleInvisibleEncoding(packagename)) mGpgExcludeInvisible else mGpg)
        if (mCore.isDbSpreadInvisibleEncoding(packagename)) {
            //spreaded doesn't work with manual padding
            res.remove(mManualZeroWidthXcoder)
        }

        return res
    }

    operator fun get(coderId: String, padderId: String?): XCoderAndPadder? {

        for (x in mAll) {
            if (x.xcoder.id == coderId) {
                if (padderId == null || x.padder == null) {
                    return x
                } else {
                    if (x.padder.id == padderId) {
                        return x
                    }
                }
            }
        }

        return null
    }

    companion object {

        @SuppressLint("StaticFieldLeak") // note that we're storing *Application*context
        @Volatile
        private var INSTANCE: XCoderAndPadderFactory? = null

        fun getInstance(ctx: Context): XCoderAndPadderFactory =
            INSTANCE ?: synchronized(this) {
                INSTANCE ?: XCoderAndPadderFactory(ctx.applicationContext).also { INSTANCE = it; it.reload() } }

    }
}